### PR TITLE
SAI-5861: optimize bitset range copy

### PR DESCRIPTION
**Pull request by magibney**
for seg-aware cache entry reconstruction

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/DocSetUtil.java [hunk 1]
```diff
@@ -335,4 +335,81 @@ is a delete (this case would trigger a bounds-checking AssertionError in
...
+   * Analogous to {@link System#arraycopy(Object, int, Object, int, int)}, but copies {@code long[]}
+   * of the format that is used to back {@link FixedBitSet}, e.g., with offset and length args
+   * specified in terms of <i>bit</i> index, not {@code long} word (array index).
+   *
+   * <p>This method is optimized to avoid inspecting individual bits, instead issuing far fewer
+   * instructions and bit-shifting "words" as necessary to align with potentially different word
+   * boundaries in the destination array. It is vastly more efficient than analogous bit-by-bit
+   * operations -- on the order of ~100x.
+   */
+  public static void copyBitRange(
```
Comment (patsonluk) [inline solr/core/src/java/org/apache/solr/search/DocSetUtil.java:357]: I get the general idea here is to cut the src range into 3 parts. first part just enough the fit the remaining "bits" of the last word of the dst indx. And then copy all the whole words (shift around based on the first cut off part), then lastly the remainder word.

I don't think it will be productive for me to read all the details and figure out if any edge case would break. 

Are there any unit test cases already to confirm this? 😁

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/search/DocSetUtil.java:357]: 🤦🏼 I must be blind or something... https://github.com/cowpaths/fullstory-solr/pull/292/files#diff-d5b9ae5bbb945a4cf6ee0f820071759cbdcd6afd06e231ff7406d701386a3654

Comment (magibney) [reply solr/core/src/java/org/apache/solr/search/DocSetUtil.java:357]: Yeah without the unit test this would be verrrry difficult to feel confident in, even if the logic seems sound. The roundtripping unit test though feels pretty solid to me. I'd almost say that the review could focus on the robustness of the test -- like, does it miss any edge cases? I _think_ the randomization should effectively cover all edge cases, but def curious if you spot any oversights!

**Review by patsonluk — APPROVED** [review]
LGTM!
